### ISSUE TITLE
Missing XML time property breaks Jenkins detection of test failures

### ISSUE THREAD

**Issue by jamesdh** [juhovh/mocha-jenkins-reporter#65]
We started noticing Jenkins incorrectly reporting all tests passing when indeed we had a couple tests failing. It was sporadic. Sometimes it would correctly identify and report them, sometimes it wouldn't. 

After a fair bit of digging around, I discovered that the XML generated by mocha-jenkins-reporter would occasionally omit the `time` property from the `<testcase>` element. This appears to happen if the given test takes less than 1/1000th of a second to run, as the lowest property I see reported is `0.001`. 

It does not matter if the property is missing from the failed test. If it's missing from any test at all passing or otherwise, then Jenkins doesn't seem to parse it correctly. 

I'm sure this is a simple fix but wanted to open an issue just to get the authors opinion. It seems fixing it here is the easiest option regardless of what the JUnit xml spec is. If Jenkins is supposed to handle the XML without a time prop then that's technically the "right" place too fix it but undoubtably much more red tape to go through...

**Comment by jamesdh** [comment]
Incorrect XML:
`<testcase classname="SuperFast" name="Executes in less than a thousandth of a second">`

Correct XML:
`<testcase classname="KindaFast" name="Executes just slow enough" time="0.001">`

**Comment by jamesdh** [comment]
According to this XSD spec (the only one I can find): https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd

...the time property is required.

**Comment by juhovh** [comment]
You are right, time should definitely be there. I think this probably broke in https://github.com/juhovh/mocha-jenkins-reporter/pull/53 PR...

Can you try to find the following lines:

```
      if (test.duration) {
        testCase.testcase[0]['_attr'].time = test.duration/1000;
      }
```
and change it to:
```
      if (test.duration) {
        testCase.testcase[0]['_attr'].time = test.duration/1000;
      } else {
        testCase.testcase[0]['_attr'].time = '0';
      }
```
and see if that helps?

**Comment by jamesdh** [comment]
Yes, I already tried something similar in that I added a default time attribute of 0, which could then be overwritten if test.duration exists: 

```
      var testCase = {
        'testcase': [{
          _attr: {
            classname: getClassName(test, currentSuite.suite),
            name: test.title,
            time: 0.000
          }
        }]
      };

      if (test.duration) {
        testCase.testcase[0]['_attr'].time = test.duration/1000;
      }
```

and can confirm it fixes it. I'll create a PR that includes that change.

**Comment by jamesdh** [comment]
This may have been a red herring. I *think* what's actually happening is that if a <testsuite> element has a time of 0, then any failures that come after that are not reported. Still digging to confirm.

**Comment by jvah** [comment]
it's not a big deal to put the time to a very small number in that case, but it's such an ugly hack I'd prefer we're sure about that first :) I'll wait until you find out

**Comment by jamesdh** [comment]
Of course, hence the digging to confirm :)

**Comment by jamesdh** [comment]
So the scenario mentioned here ultimately ended up being the cause: 
https://github.com/jenkinsci/junit-plugin/pull/63/files#r97906666

Crazy frustrating to debug, made even more frustrating because the changes I made to test out this issue worked repeatedly, but that was just by pure coincidence. The original PR isn't a bad thing, as I guess the time attribute should be there anyways. Unfortunately that won't give me back the time I spent diagnosing this 😭 

Given the fact that mocha tests generally run much faster than traditional Java tests (what JUnit was originally written for) and that the time stamps only have second granularity, I could easily see this being a subtle issue that goes unnoticed. It might be worthwhile creating a PR that appends something to the test suite names to make them unique in the situation that multiple test suites have the same name and timestamp. Thoughts?
